Implement: Priorities

fran-cap opened the pull request:
Adds the priorities panel used to set ai tasks.
Created constants for inc/dec and max priorities values.
Changed data types of priority list type and name to match decompiliation (double check this).

Implements:
	LegoRR::Priorities_LoadImages
	LegoRR::Priorities_Reset
	LegoRR::Priorities_LoadLevel
	LegoRR::Priorities_LoadPositions
	LegoRR::Priorities_HandleInput
	LegoRR::Priorities_ChangePriorityPressed
	LegoRR::Priorities_SetHoverArea
	LegoRR::Priorities_MovePriorityUpOrTop
	LegoRR::Priorities_MovePriorityTop
	LegoRR::Priorities_MovePriorityUp
	LegoRR::Priorities_TurnPriorityOff
	LegoRR::Priorities_MoveCursorToPriorityUpButton
	LegoRR::Priorities_MoveCursorToPriorityImage
	LegoRR::Priorities_Draw

fran-cap committed:
Implement: Priorities

Adds the priorities panel used to set ai tasks.
Created constants for inc/dec and max priorities values
Changed data names of priority list type and name to match decompiliation (double check this).

Implements:
	LegoRR::Priorities_LoadImages
	LegoRR::Priorities_Reset
	LegoRR::Priorities_LoadLevel
	LegoRR::Priorities_LoadPositions
	LegoRR::Priorities_HandleInput
	LegoRR::Priorities_ChangePriorityPressed
	LegoRR::Priorities_SetHoverArea
	LegoRR::Priorities_MovePriorityUpOrTop
	LegoRR::Priorities_MovePriorityTop
	LegoRR::Priorities_MovePriorityUp
	LegoRR::Priorities_TurnPriorityOff
	LegoRR::Priorities_MoveCursorToPriorityUpButton
	LegoRR::Priorities_MoveCursorToPriorityImage
	LegoRR::Priorities_Draw

trigger-segfault reviewed (commented):
Good work. Some changes need to be made, and there's one bug (which is actually my fault because it was like that in the comments I wrote).

fran-cap commented on src/openlrr/game/interface/Priorities.cpp:
```diff
@@ -28,45 +39,365 @@ LegoRR::Priorities_Globs & LegoRR::prioritiesGlobs = *(LegoRR::Priorities_Globs*
+
+	for(uint32 i = 0; i < AI_Priority_Count; i++){
+		orderIndex++;
+		_itoa(orderIndex, (char*)&gameName, 10);
```
Small note here, due to it being orderIndex++; we need to do a +1 on i. Also VS throws an error for itoa not being part of std so I kept ht _itoa:

```C++
char orderName[12]; // Max number of characters +1 that can appear in signed int32.
_itoa((int)i+1, orderName, 10);
```

fran-cap committed:
Fix #79: Code cleanup and (isOff == 0) != 0 bug

Cleaned up a lot of == 0 errors and casts to (char*).
Fixed bug where the warning message for ore disabling was inverted.
Logic cleaned up for priority assignments.

fran-cap commented:
Ok this should capture the above changes.
Thanks for verifying that I'm not crazy with that disable message ;).

trigger-segfault commented on src/openlrr/game/interface/Priorities.cpp:
```diff
@@ -28,45 +39,361 @@ LegoRR::Priorities_Globs & LegoRR::prioritiesGlobs = *(LegoRR::Priorities_Globs*
+	char buff[512];
+	
+	const char* arrayID = Config_ID(gameName, "PriorityImages");
+	for (const Gods98::Config* prop = Config_FindArray(config, arrayID); prop != nullptr;  prop = (Gods98::Config*)Config_GetNextItem(prop)) {
```
`(Gods98::Config*)` cast needs to be removed for `Config_GetNextItem`.

trigger-segfault commented on src/openlrr/game/interface/Priorities.cpp:
```diff
@@ -28,45 +39,361 @@ LegoRR::Priorities_Globs & LegoRR::prioritiesGlobs = *(LegoRR::Priorities_Globs*
+	char* arrayID;
+	const Gods98::Config* prop;
+	prioritiesGlobs.count = 0;
+	arrayID = (char*)Config_ID(gameName, levelName, "Priorities");
```
`arrayID` needs to be const and `(char*)` cast should be removed.

trigger-segfault commented on src/openlrr/game/interface/Priorities.cpp:
```diff
@@ -28,45 +39,361 @@ LegoRR::Priorities_Globs & LegoRR::prioritiesGlobs = *(LegoRR::Priorities_Globs*
+	for (uint32 i = 0; i < AI_Priority_Count; i++) {
+		char orderName[12]; // Max number of characters +1 that can appear in signed int32.
+		_itoa((int)i+1, orderName, 10);
+		keyPath = (char*)Config_ID(startGameName, "PrioritiesImagePositions", &orderName);
```
`keyPath` needs to be const and `(char*)` cast should be removed.

`startGameName` isn't needed anymore, so pass `gameName` instead and remove the local variable.

trigger-segfault commented on src/openlrr/game/interface/Priorities.cpp:
```diff
@@ -28,45 +39,361 @@ LegoRR::Priorities_Globs & LegoRR::prioritiesGlobs = *(LegoRR::Priorities_Globs*
+				//priorityType = (AI_Priority)prioritiesGlobs.sfxPriorityName[index + 0x1a];
+				priorityType = prioritiesGlobs.buttonTypes[index-1];
+				prioritiesGlobs.buttonTypes[index] = priorityType;
+				AITask_Game_SetPriorityValue(priorityType, MAX_PRIO + (index * -PRIO_INC));
```
This area is one tab level too high.

trigger-segfault reviewed (commented):
Heh yeah, I remember questioning myself with the isOff section back when I was commenting that in Ghidra. Good thing the result of that condition was easy to identify, or this would have gone unnoticed.

Good catch with the 1-indexed order index for itoa. :)

A few more changes. And the `orderIndex > 0` is still missing from MovePriorityUp.

fran-cap committed:
Fix #79: Add missing > 0 check, remove unneeded casts

Also had some extra whitespace removed in MovePriorityTop

fran-cap commented:
Thanks for the 2nd review. Pushed up newest changes.

trigger-segfault commented:
Everything looks good. Thanks for your work. :)
